pkg/relayer/proxy/relay_meter.go

```diff
@@ -0,0 +1,327 @@
+	sharedtypes "github.com/pokt-network/poktroll/x/shared/types"
+)
+
+var _ relayer.RelayMeter = (*ProxyRelayMeter)(nil)
+
+// appRelayMeter is the relay meter's internal representation of an application's
+// max and consumed stake.
+type appRelayMeter struct {
+	maxAmount      cosmostypes.Coin
+	consumedAmount cosmostypes.Coin
+	app            apptypes.Application
+}
+
+// ProxyRelayMeter is the off-chain Supplier's rate limiter.
+// It ensures that no application is over-serviced by the Supplier by maintaining
+// the max amount of stake the supplier can consume per session and the amount of
+// stake consumed by mined relays.
+type ProxyRelayMeter struct {
+	// The known applications that have their stakes metered.
+	apps map[string]*appRelayMeter
+	// overServicingAllowance adjusts the max amount to allow for controlled over-servicing
+	// or, if negative, to create a stake buffer.
+	// TODO_TECHDEBT(@red-0ne): Expose overServicingAllowance as a configuration parameter.
+	overServicingAllowance int64
+
+	applicationQuerier client.ApplicationQueryClient
+	serviceQuerier     client.ServiceQueryClient
+	sharedQuerier      client.SharedQueryClient
+	eventsQueryClient  client.EventsQueryClient
+	blockQuerier       client.BlockClient
+
+	relayMeterMu sync.Mutex
+}
+
+func NewRelayMeter(deps depinject.Config) (relayer.RelayMeter, error) {
+	rm := &ProxyRelayMeter{
+		apps:                   make(map[string]*appRelayMeter),
+		overServicingAllowance: 1000000,
+	}
+
+	if err := depinject.Inject(
+		deps,
+		&rm.sharedQuerier,
+		&rm.applicationQuerier,
+		&rm.serviceQuerier,
+		&rm.blockQuerier,
+		&rm.eventsQueryClient,
+	); err != nil {
+		return nil, err
+	}
+
+	return rm, nil
+}
+
+// Start starts the relay meter by observing application staked events and new sessions.
+func (rmtr *ProxyRelayMeter) Start(ctx context.Context) error {
+	eventsObs, err := rmtr.eventsQueryClient.EventsBytes(ctx, "tm.event = 'Tx'")
+	if err != nil {
+		return err
+	}
+
+	// Listen to application staked events and update known application stakes.
+	appStakedEvents := filterTypedEvents[*apptypes.EventApplicationStaked](ctx, eventsObs, nil)
+	channel.ForEach(ctx, appStakedEvents, rmtr.forEachEventApplicationStakedFn)
+
+	// Listen to new blocks and reset the relay meter application stakes every new session.
+	committedBlocksSequence := rmtr.blockQuerier.CommittedBlocksSequence(ctx)
+	channel.ForEach(ctx, committedBlocksSequence, rmtr.forEachNewBlockFn)
+
+	return nil
+}
+
+// ClaimRelayPrice claims the relay price for the given relay request metadata.
+// It deducts the relay cost from the application's stake and returns an error if
+// the application has been rate limited.
+func (rmtr *ProxyRelayMeter) ClaimRelayPrice(ctx context.Context, reqMeta servicetypes.RelayRequestMetadata) error {
+	rmtr.relayMeterMu.Lock()
+	defer rmtr.relayMeterMu.Unlock()
+
+	sharedParams, err := rmtr.sharedQuerier.GetParams(ctx)
```

Comment:
This function does A LOT of onchain querying.
- What if we start handling 1000s of requests per second?
- What if we start sending those requests to another relayMiner beside a pocket node?
- This will exponentiate quickly...

1. Does it get cached? If so, how. #PUC
2. If not, how will cache it? #PUC and add a TODO_BETA

Reply:
It is currently not cached but we definitely need to cache all of these. But is definitely on my radar (need it for load testing too)

We have a #543 to add caching mechanism with granular invalidation to all our query clients.

I will update the code so those are only fetched for a new apps. Since `appToRelayMeterMap` is emptied at each new session.